File: Best Time to Buy and Sell Stock.cpp
/*
扫描一遍，记录下遇到的min price，拿当前price减去min price，
记录下最大的profit
*/
class Solution {
public :
    int maxProfit(vector< int> &prices) {
        int min_price, profit = 0, len = prices.size();
        if(len) min_price = prices[0]; //初始化最小价格
        for(int i=1; i<len; i++){
			//更新最小价格
			min_price = min(prices[i], min_price);
			//用当前最小价格计算最大利润
            profit = prices[i]-min_price > profit? prices[i]-min_price : profit;
        }
        return profit;
    }
};

File: python/Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if len(prices) == 0:
            return 0
        minPrice = prices[0]
        profit = 0
        for i in xrange(1,len(prices)):
            if prices[i] < minPrice:
                minPrice = prices[i]
            if prices[i]-minPrice > profit:
                profit = prices[i]-minPrice
        return profit

#test part
S = Solution()
Prices = [6,1,3,2,4,7]
print S.maxProfit(Prices)